examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+            let input_ty_skipped_binder: &ty::Ty = input_ty.skip_binder(); // Skip binder once
+            if let ty::TyKind::Param(param_ty) = input_ty_skipped_binder.kind() {
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter is bound only by Iterator
+                if is_param_bound_only_by_iterator(
+                    cx,
+                    predicates,
+                    *input_ty_skipped_binder,
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) && !is_param_used_in_other_trait_args(
+                    cx,
+                    predicates,
+                    *input_ty_skipped_binder,
+                    iterator_def_id,
+                    sized_def_id,
+                    id,
+                ) {
+                    let input_hir_ty = &decl.inputs[i];
+                    span_lint_and_help(
+                        cx,
+                        ARG_ITER,
+                        input_hir_ty.span,
+                        "parameter type has `Iterator` bound",
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Checks if a given type parameter `param_ty` is bound _only_ by `iterator_def_id`
+/// within the given `predicates`
+fn is_param_bound_only_by_iterator<'tcx>(
+    _cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_ty: ty::Ty<'tcx>,
+    iterator_def_id: DefId,
+    sized_def_id: DefId,
+    _fn_id: LocalDefId,
+) -> bool {
+    let mut found_iterator_bound = false;
+
+    for &(predicate, _pred_span) in predicates.predicates.iter() {
+        if let Some(trait_pred) = predicate.as_trait_clause() {
+            let bound_trait_pred = trait_pred.skip_binder();
+            let self_ty = bound_trait_pred.self_ty();
+
+            if self_ty == param_ty {
+                let current_trait_def_id = bound_trait_pred.def_id();
+
+                if current_trait_def_id == iterator_def_id {
+                    found_iterator_bound = true;
+                } else if current_trait_def_id != sized_def_id {
+                    return false;
+                }
+            }
+        }
+    }
+
+    found_iterator_bound
+}
+
+/// Checks if the given parameter type `param_ty` is used as an argument
+fn is_param_used_in_other_trait_args<'tcx>(
+    _cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_ty: ty::Ty<'tcx>,
+    iterator_def_id: DefId,
+    sized_def_id: DefId,
+    _fn_id: LocalDefId,
+) -> bool {
+    if !matches!(param_ty.kind(), ty::TyKind::Param(_)) {
+        return false;
+    }
```

Comment:
```suggestion
```
Unless you feel strongly, I would drop this. It's verified that `param_ty` is a `Parm` just before `is_param_used_in_other_trait_args` is called. I don't think it's necessary to check it again here.